src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT = 10;
+    private static final double STANDART_GROW = 1.5;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT];
+    }
+
     @Override
     public void add(T value) {
-
+        capacityEnsures();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        addingChecking(index);
+        capacityEnsures();
+        for (int i = size - 1; i >= index; i--) {
+            elements[i + 1] = elements[i];
+        }
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        capacityChecking(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        capacityChecking(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        capacityChecking(index);
+        final T temp = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return temp;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] == element)
+                    || (elements[i] != null
+                    && elements[i].equals(element))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Can't found element by value " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void capacityEnsures() {
+        if (size == elements.length) {
+            int capacity = (int) (elements.length + STANDART_GROW);
+            T[] tempArray = (T[]) new Object[capacity];
+            System.arraycopy(elements, 0, tempArray, 0, size);
+            elements = tempArray;
+        }
+    }
+
+    private void capacityChecking(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("This index out of bounds " + index);
+        }
+    }
+
+    private void addingChecking(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on position " + index);
+        }
     }
```

Comment:
These methods are the same, read comment to the previous PR

Reply:
So in one **index > size** but in another **index >= size**
it is not one and the same, no?

Reply:
Answered in your channel.